src/coreclr/jit/lower.cpp

```diff
@@ -3777,12 +3777,47 @@ void Lowering::LowerRetStruct(GenTreeUnOp* ret)
 
         case GT_BLK:
         case GT_OBJ:
-            retVal->ChangeOper(GT_IND);
-            FALLTHROUGH;
         case GT_IND:
+        {
+            // Spill to a local if sizes don't match so we can avoid the "load more than requested"
+            // problem, e.g. struct size is 5 and we emit "ldr x0, [x1]"
+            unsigned realSize = retVal->OperIs(GT_OBJ) ? retVal->AsObj()->Size() : retVal->AsIndir()->Size();
+            CORINFO_CLASS_HANDLE structCls = NO_CLASS_HANDLE;
+            if (realSize == 0)
+            {
+                assert(retVal->gtNext->OperIs(GT_RETURN));
+
+                // We have an IND<struct> node, assuming it's used in a return statement, take target size
+                // from the current method's signature (it's only used to decide whether we need to spill it to
+                // a local or not, so the actual size won't be used).
+                // TODO-ADDR: delete once "IND<struct>" nodes are no more
+                structCls = comp->info.compMethodInfo->args.retTypeClass;
+                realSize  = comp->info.compCompHnd->getClassSize(structCls);
+            }
+
+            if (genTypeSize(nativeReturnType) > realSize)
+            {
+                LIR::Use retValUse(BlockRange(), &ret->gtOp1, ret);
+                unsigned tmpNum = BAD_VAR_NUM;
+                if (structCls != NO_CLASS_HANDLE)
+                {
+                    tmpNum               = comp->lvaGrabTemp(true DEBUGARG("spilling IND(X)"));
+                    comp->genReturnLocal = tmpNum;
+                    comp->lvaSetStruct(tmpNum, structCls, true);
+                }
+                ReplaceWithLclVar(retValUse, tmpNum);
+                LowerRetSingleRegStructLclVar(ret);
+                break;
+            }
+
+            if (retVal->OperIs(GT_OBJ))
+            {
+                retVal->ChangeOper(GT_IND);
+            }
```

Comment:
```suggestion
            retVal->ChangeOper(GT_IND);
```
Should work as it did before? In any case, `BLK`s also need to be turned into `IND`s.